src/core/lib/transport/metadata.cc

```diff
@@ -365,14 +404,101 @@ grpc_mdelem grpc_mdelem_create(
   return GRPC_MAKE_MDELEM(md, GRPC_MDELEM_STORAGE_INTERNED);
 }
 
-grpc_mdelem grpc_mdelem_from_slices(const grpc_slice& key,
-                                    const grpc_slice& value) {
-  grpc_mdelem out = grpc_mdelem_create(key, value, nullptr);
-  grpc_slice_unref_internal(key);
+template <bool key_definitely_static>
+static grpc_mdelem md_create_maybe_static(const grpc_slice& key,
+                                          const grpc_slice& value) {
+  const bool key_is_static_mdstr =
+      key_definitely_static || GRPC_IS_STATIC_METADATA_STRING(key);
+  const intptr_t kidx = GRPC_STATIC_METADATA_INDEX(key);
+
+  // Not all static slice input yields a statically stored metadata element.
+  // It may be worth documenting why.
+  if (key_is_static_mdstr && GRPC_IS_STATIC_METADATA_STRING(value)) {
+    grpc_mdelem static_elem = grpc_static_mdelem_for_static_strings(
+        kidx, GRPC_STATIC_METADATA_INDEX(value));
+    if (!GRPC_MDISNULL(static_elem)) {
+      return static_elem;
+    }
+  }
+
+  uint32_t khash = key_definitely_static
+                       ? grpc_static_metadata_hash_values[kidx]
+                       : grpc_slice_hash_refcounted(key);
+
+  uint32_t hash = GRPC_MDSTR_KV_HASH(khash, grpc_slice_hash_refcounted(value));
+  return md_create_must_intern<key_definitely_static>(key, value, hash);
+}
+
+grpc_mdelem grpc_mdelem_create_static_key_interned_val(
+    const grpc_slice& key, const grpc_slice& value) {
+  grpc_mdelem ret = md_create_maybe_static<true>(key, value);
+  grpc_slice_unref_internal(value);
+  return ret;
+}
+
+template <bool key_definitely_static>
+static grpc_mdelem md_create(
```

Comment:
Could you add some comments above each of these new functions explaining when we're supposed to use them?

Reply:
Will do.

Reply:
Updated PR has comments, but more importantly, the programmer need no longer actively think about which method to pick due to the strongly typed slices.